timemachine/fe/absolute_hydration.py

```diff
@@ -0,0 +1,142 @@
+    """solvent + (noninteracting ligand) sample --> solvent + (vacuum ligand) sample
+
+    Inputs
+    ------
+    * solvent_samples ~ p_noninteracting([x_solvent, x_ligand])
+    * (ligand_samples, ligand_log_weights) ~ p_vacuum(x_ligand)
+        (note: this set of samples is importance-weighted!)
+
+    Processing
+    ----------
+    * resample ligand_samples according to ligand_log_weights
+    * concatenate solvent component from p_noninteracting with ligand from p_vacuum
+
+
+    Returns
+    -------
+    * list of (coordinates, velocities, box), distributed according to
+
+
+    Assumptions
+    -----------
+    * ligand indices: last num_ligand_atoms
+
+    Notes
+    -----
+    * TODO[generality]: refactor to accept two streams of unweighted samples, concatenate them
+        (rather than requiring and discarding ligand component in solvent_samples, performing multinomial resampling)
+    """
+
+    # assume this layout
+    num_total_atoms = len(solvent_samples[0].coords)
+    num_solvent_atoms = num_total_atoms - num_ligand_atoms
+    solvent_idxs = np.arange(0, num_solvent_atoms)
+
+    # sample according to log weights
+    ligand_xvs = enhanced.sample_from_log_weights(ligand_samples, ligand_log_weights, size=num_samples)
+
+    # sample uniformly with replacement
+    solvent_choice_idxs = np.random.choice(np.arange(len(solvent_samples)), size=num_samples)
+
+    all_xvbs = []
+    for i, choice_idx in enumerate(solvent_choice_idxs):
+
+        # solvent + noninteracting ligand
+        noninteracting_xvb = solvent_samples[choice_idx]
+
+        # vacuum ligand
+        ligand_x, ligand_v = ligand_xvs[i]
+
+        # concatenate solvent
+        combined_x = np.concatenate([noninteracting_xvb.coords[solvent_idxs], ligand_x], axis=0)
+        combined_v = np.concatenate([noninteracting_xvb.velocities[solvent_idxs], ligand_v], axis=0)
+
+        combined_xvb = CoordsVelBox(combined_x, combined_v, noninteracting_xvb.box)
+
+        all_xvbs.append(combined_xvb)
+    return all_xvbs
+
+
+def setup_absolute_hydration_with_endpoint_samples(mol, temperature=300.0, pressure=1.0, n_steps=1000, seed=2022):
+    """Generate samples from the equilibrium distribution at lambda=1
+
+    Return:
+    * reduced_potential_fxn
+    * npt_mover
+    * initial_samples from lam = 1
+    """
+    if type(seed) != int:
+        seed = np.random.randint(1000)
+        print(f"setting seed randomly to {seed}")
+    else:
+        print(f"setting seed to {seed}")
+
+    np.random.seed(seed)
+
+    # set up potentials
+    ff = Forcefield.load_from_file("smirnoff_1_1_0_ccc.py")
+    potentials, params, masses, _, _ = enhanced.get_solvent_phase_system(mol, ff)
+    potential_fxn = construct_potential(potentials, params)
+
+    kBT = BOLTZ * temperature
+
+    def reduced_potential_fxn(xvb, lam):
+        return potential_fxn(xvb, lam) / kBT
+
+    bind_potentials(potentials, params)
+
+    # set up npt mover
```

Comment:
nit: redundant comment

Reply:
Removed in https://github.com/proteneer/timemachine/pull/718/commits/3a99dc756c6701dc9e685c6d123c3f378c7e7e3e